Fix SAML add/remove actions in the Account Security page

miveronese opened the pull request:
Issue: #43256

This PR adds support for SAML on the Account Security page and fixes the add/remove actions.

## Test plan
All tests should pass.
Manually tested locally as shown in the attached video.

https://user-images.githubusercontent.com/1552863/197095501-8db9a086-e6ad-4bd1-92bb-2539d28b0acc.mov


<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles -->

## App preview:

- [Web](https://sg-web-mv-bugfix-account-security-page.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

sg-e2e-regression-test-bob commented:
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (+0.13 kb) | 0.03% (+3.90 kb) | 0.03% (+3.77 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/mv/bugfix-account-security-page/compare-report.html;tab=live_object) for a full comparison between the commits 1f0583e and 838a9cd or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

mrnugget commented on client/web/src/user/settings/auth/ExternalAccount.tsx:
```diff
@@ -52,16 +57,24 @@ export const ExternalAccount: React.FunctionComponent<React.PropsWithChildren<Pr
                 <div className="text-muted">
-                    {account.external ? (
+                    {authProvider.serviceType === 'saml' && (
+                        <>{account.external ? <>{account.external.userName}</> : 'Not connected'}</>
```
```suggestion
                        {account.external ? <>{account.external.userName}</> : 'Not connected'}
```

Are these needed?

miveronese replied:
Yes -- without them, the page won't load and you'll see a huge "module build failed..." error

mrnugget commented on client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx:
```diff
@@ -26,15 +25,37 @@ interface GitLabExternalData {
         userName: string
-        userLogin: string
-        userUrl: string
+        userLogin: string | null
+        userUrl: string | null
```
```suggestion
        userLogin?: string
        userUrl?: string
```

I'm not 100% sure but I think this would be more the convention in our codebase.

kopancek replied:
+1

mrnugget commented on client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx:
```diff
@@ -50,12 +71,12 @@ const getNormalizedAccount = (
 ): NormalizedMinAccount => {
     // kind and type match except for the casing
-    const kind = authProvider.serviceType.toLocaleUpperCase() as ExternalServiceKind
+    const kind = authProvider.serviceType.toLocaleUpperCase() as ExternalAccountKind
```
What happens when the cast fails? When I hover over `serviceType` I see the following:

```
serviceType: "github" | "gitlab" | "http-header" | "openidconnect" | "saml" | "builtin"
```

But we only have 3 `ExternalAccountKind` defined.

Should we maybe do a switch-statement based casting? i.e. `if authProvider.servicetype == "github" { kind = Externalaccountkind.github }` (but in a better way 😄 )

kopancek replied:
I'm surprised that TS did compile on this, but instead the TS build fails on the `null` type definitions now: https://buildkite.com/sourcegraph/sourcegraph/builds/179151#0183f85e-b8fb-4ece-9dea-3f4358c6c350 

I agree with @mrnugget above, this seems like unsafe casting and we should guard it somehow.

miveronese replied:
The null type is fixed.
I will look into the suggestions for the casting.

Just a note:
- previously, we had `const kind = authProvider.serviceType.toLocaleUpperCase() as ExternalServiceKind`
- and  the  kinds defined in `ExternalServiceKind` didn't totally match either `serviceType: "github" | "gitlab" | "http-header" | "openidconnect" | "saml" | "builtin"` 
- was that something that should not have been compiled (that was wrong) either? 🤔 I used the previous implementation as a reference and this question comes from my lack of TS knowledge :D

mrnugget commented on client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx:
```diff
@@ -95,6 +116,33 @@ const getNormalizedAccount = (
+            case 'SAML':
+                {
+                    const samlExternalData = accountExternalData as SamlExternalData
+                    const data = samlExternalData.Assertions[0].AttributeStatement.Attributes
```
Can we assume that `.Assertions` always has `.length >= 1`? What happens if it's `0`?

miveronese replied:
Good point.
before our backend "builds" the SAML account_data that is going to be stored in the DB (a flow triggered after the user clicks the Add button on this page), we do all the validations necessary, including checking if Assertions length = `0`.
What  can be improved here:
- making sure nothing will "explode"
- surface to the UI, in a friendly way, the errors that come from the backend.

Let me know if you agree.

miveronese replied:
This is how the error will look like
https://github.com/sourcegraph/sourcegraph/pull/43274#discussion_r1003757301

mrnugget commented on client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx:
```diff
@@ -95,6 +116,33 @@ const getNormalizedAccount = (
+                    const data = samlExternalData.Assertions[0].AttributeStatement.Attributes
+                    let nickname = ''
+
+                    for (let index = 0, length = data.length; index < length; index += 1) {
+                        const item = data[index]
+                        if (data[index].Values && item.Name) {
```
I think we prefer [`for ... of`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of) whenever possible.

I think this should work:

```suggestion
                    for (const item of data) {
                        if (item.Values && item.Name) {
```

kopancek replied:
+1

mrnugget commented on client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx:
```diff
@@ -95,6 +116,33 @@ const getNormalizedAccount = (
+                        if (data[index].Values && item.Name) {
+                            const key = item.Name.slice(Math.max(0, item.Name.lastIndexOf('/') + 1))
+                            if (key === 'nickname') {
+                                nickname = item.Values[0].Value
```
There's no `break` after this, which means that if `data` contains multiple `item`s with that `key` then each one will assign to `nickname` and the last one is used. Is that the intention?

kopancek replied:
Also, can we safely assume that `Values` is always length > 0? If it's an empty array, this code would fail.

miveronese replied:
Please, see the recent changes on how this part was handled and refactored.

@mrnugget , I added a break after we get the username now (if the data doesn't have a username -- this can happen --, then I attempt to get an email).

@kopancek, if we don't get, from our backend, any values (`Values` length === 0), we won't even get to this part of the code/page. 
Before saving the saml information to the user_external_accounts > account_data, our backend tries  to get the values from a SAML assertion (from which the email will be extracted). If there are no values, our backend returns an error during the redirect and the user will see something like this

<img width="1010" alt="Screen Shot 2022-10-24 at 2 51 27 PM" src="https://user-images.githubusercontent.com/1552863/197626971-c438ad64-f68e-428e-9fe5-cd07c54f6b44.png">
.

mrnugget commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -2642,6 +2642,15 @@ type ExternalServiceSyncJob implements Node {
 }
 
+"""
+A specific kind of external account.
```
I'd add a bit more information here:
```suggestion
The different kinds of external account that can be connected to a Sourcegraph user.
```

mrnugget commented on client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx:
```diff
@@ -95,6 +116,33 @@ const getNormalizedAccount = (
+                    for (let index = 0, length = data.length; index < length; index += 1) {
+                        const item = data[index]
+                        if (data[index].Values && item.Name) {
+                            const key = item.Name.slice(Math.max(0, item.Name.lastIndexOf('/') + 1))
```
I think either a comment that explains how the `item.Name` is transformed or a test would go a long way here 😄

kopancek replied:
I think we can write this as:
```suggestion
                            const key = item.Name.split('/').pop()
```

Which should make the code more readable and self-explanatory (I hope 😄 )

kopancek commented on client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx:
```diff
@@ -95,6 +116,33 @@ const getNormalizedAccount = (
+                        external: {
+                            id: account.id,
+                            userName: nickname,
+                            userLogin: null,
```
I think we can also safely assume:
```suggestion
                            userLogin: nickname,
```

kopancek commented on cmd/frontend/graphqlbackend/schema.graphql:
```diff
@@ -2642,6 +2642,15 @@ type ExternalServiceSyncJob implements Node {
+"""
+A specific kind of external account.
+"""
+enum ExternalAccountKind {
+    GITHUB
+    GITLAB
+    SAML
+}
```
Do we really need this as part of the graphql schema? I don't see it used anywhere in the graphql schema itself. If we need a definition on the UI for typescript benefit, let's define one in typescript, e.g. by "inheriting" the `AuthProvider.serviceType` and omitting `http-header` and `builtin` auth. Something like using [Exclude](https://www.typescriptlang.org/docs/handbook/utility-types.html#excludeuniontype-excludedmembers):
```
type ExternalAccountKind = Exclude<AuthProvider.serviceType, "http-header" | "builtin">
```

mrnugget replied:
+1 here. I don't think we need to define this in the GraphQL schema (although the idea of only using the GraphQL schema for type definitions does intrigue me)

kopancek commented on client/web/src/components/externalAccounts/externalAccounts.tsx:
```diff
@@ -0,0 +1,46 @@
+
+import { ExternalAccountKind } from '@sourcegraph/shared/src/graphql-operations';
+
+export interface AddExternalAccountOptions {
```
```suggestion
export interface ExternalAccount {
```

kopancek commented on client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx:
```diff
@@ -72,7 +93,7 @@ const getNormalizedAccount = (
                         external: {
                             id: account.id,
-                            // map github fields
+                            // map GitHub fields
```
is this comment change needed?

miveronese replied:
my IDE complained that we use "GitHub" instead of "github" in the rest of the codebase. so why not change it? 😂

kopancek commented on client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx:
```diff
@@ -95,6 +116,33 @@ const getNormalizedAccount = (
+                        const item = data[index]
+                        if (data[index].Values && item.Name) {
+                            const key = item.Name.slice(Math.max(0, item.Name.lastIndexOf('/') + 1))
+                            if (key === 'nickname') {
```
Where does `'nickname'` come from? Can you point me to a SAML specification that says this attribute is always present and contains the name?

miveronese replied:
Good question! As `nickname` in the example, I used as a reference (Auth0) is part of the Attribute Statements and Attribute Statements are optional, this specific field is set accordingly to the will of the IdP, hence it's not only present. See the recent changes where I am following what we are doing in the backend to get the username.

miveronese commented on client/web/src/user/settings/auth/ExternalAccount.tsx:
```diff
@@ -52,16 +57,24 @@ export const ExternalAccount: React.FunctionComponent<React.PropsWithChildren<Pr
                 <H3 className="m-0">{authProvider.displayName}</H3>
                 <div className="text-muted">
-                    {account.external ? (
+                    {authProvider.serviceType === 'saml' && (
+                        <>{account.external ? <>{account.external.userName}</> : 'Not connected'}</>
+                    )}
```
1. Removing the tags as you suggested won't work -- it will trigger a `Module build failed` error.

2. For gitlab and github, we have a link to the user page in the external service (such as https://github.com/kopancek). For SAML, we don't have that link/page, that's why I'm skipping it.

<img width="883" alt="Screen Shot 2022-10-21 at 10 20 40 AM" src="https://user-images.githubusercontent.com/1552863/197242831-d8853dbb-b91c-427b-97c2-8d2964b38888.png">

kopancek replied:
This will definitely work without the React.fragment `<>` shenanigans:
```javascript
                    {authProvider.serviceType === 'saml' &&
                        (account.external ? account.external.userName : 'Not connected')}
                    {authProvider.serviceType !== 'saml' && (
                        <>
                            {account.external?.userUrl ? (
                                <>
                                    {account.external.userName} (
                                    <Link to={account.external.userUrl} target="_blank" rel="noopener noreferrer">
                                        @{account.external.userLogin}
                                    </Link>
                                    )
                                </>
                            ) : (
                                'Not connected'
                            )}
                        </>
                    )}
```

miveronese replied:
Ha!! Cool!

miveronese commented on client/web/src/user/settings/auth/ExternalAccount.tsx:
```diff
@@ -30,8 +30,13 @@ export const ExternalAccount: React.FunctionComponent<React.PropsWithChildren<Pr
-    }, [authProvider.authenticationURL])
+
+        if (authProvider.serviceType === 'saml') {
+            window.location.assign(`${authProvider.authenticationURL}`)
```
If we also add `redirect=${window.location.href}` to the `saml` case:

- the redirect URL will be built in the  wrong way
-  it will look like this (example from a local instance) `https://sourcegraph.test:3443/.auth/saml/login&redirect=https:/sourcegraph.test:3443/users/<USERNAME>/settings/security` 
- and we will have a ton of failing attempts to redirect

miveronese commented on client/web/src/components/externalAccounts/externalAccounts.tsx:
```diff
@@ -0,0 +1,46 @@
+    kind: ExternalAccountKind.SAML,
+    title: 'SAML',
+    icon: AccountCircleIcon,
+}
```
> What about OpenIDConnect?
As we currently don't offer support for it the Acct Security Page, I've chosen to divide the task into parts, starting first with adding support to SAML.
But that shouldn't be a blocker to adding, in this PR, the other auth options you suggested and updating their naming :)

miveronese commented:
converting this PR to a Draft mode while working on the udpates.

mrnugget commented:
> converting this PR to a Draft mode while working on the udpates.

Ping when ready for re-review :)

miveronese commented on client/web/src/components/externalAccounts/externalAccounts.tsx:
```diff
@@ -0,0 +1,54 @@
+    ['gitlab']: GITLAB,
+    ['saml']: SAML,
+    ['openidconnect']:OPENID_CONNECT,
+}
```
Is this acceptable in TS?

pjlast replied:
```suggestion
export const defaultExternalAccounts: Record<ExternalAccountKind, ExternalAccount> = {
    github: GITHUB,
    gitlab: GITLAB,
    saml: SAML,
    openidconnect: OPENID_CONNECT,
```

You can do it like this, neither the square brackets or the quotes are necessary. It's a JavaScript object, so it will treat the keys here as strings, just like the objects above it

miveronese commented on client/web/src/components/externalAccounts/externalAccounts.tsx:
```diff
@@ -0,0 +1,46 @@
+    [ExternalAccountKind.GITHUB]: GITHUB_DOTCOM,
+    [ExternalAccountKind.GITLAB]: GITLAB_DOTCOM,
+    [ExternalAccountKind.SAML]: SAML,
+}
```
I could not find a way to make it work without keeping this piece of Record 🤔

miveronese commented on client/web/src/user/settings/auth/ExternalAccount.tsx:
```diff
@@ -30,8 +30,13 @@ export const ExternalAccount: React.FunctionComponent<React.PropsWithChildren<Pr
-        window.location.assign(`${authProvider.authenticationURL}&redirect=${window.location.href}`)
-    }, [authProvider.authenticationURL])
+
+        if (authProvider.serviceType === 'saml') {
+            window.location.assign(`${authProvider.authenticationURL}`)
+        } else {
+            window.location.assign(`${authProvider.authenticationURL}&redirect=${window.location.href}`)
```
It looks like it won't work using history here as the navigation will happen only on the client side and we need the app here to access the server side so the saml flow will be complete

kopancek replied:
You're right, `history` navigation does not work well with external URLs. Let's keep using the `location.assign`

pjlast commented on client/web/src/components/externalAccounts/externalAccounts.tsx:
```diff
@@ -0,0 +1,54 @@
+import React from 'react';
```
This file doesn't seem to have any embedded HTML/React components, it's just helper types, so it should probably just be a `.ts` file and not a `.tsx` file.

pjlast commented on client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx:
```diff
@@ -65,35 +79,63 @@ const getNormalizedAccount = (
+            }
+                break
+            case 'saml': {
+                // The data stored in user_external_account has this structure: “Values”[“<attribute names for email or username>"] > Values[0] > “Value”: string
```
```suggestion
                // The data stored in user_external_account has this structure: "Values"["<attribute names for email or username>"] > Values[0] > "Value": string
```

There's a mix of different kinds of quotes here

mrnugget commented on client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx:
```diff
@@ -65,35 +79,63 @@ const getNormalizedAccount = (
+                        if (usernameAttrs.has(name)) {
+                            username = att.Values.length > 0 && att.Values[0].Value || ''
+
+                            break
```
That means once we find the `username` we stop iterating over `entries` -- what if `email` hasn't been set yet? Is that okay? Why no `break` in the other conditional?

I know it's nit-picky, but: would be great to to pull line 115-128 out into a function and write a few short unit tests for it. Unit tests for pure TS are pretty easy to write too. See [this file](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/client/web/src/util/url.test.ts) for examples of small tests for utility functions.

miveronese replied:
Good idea (about pulling that part out so we can test it) and thanks for the example!
On the email/break part, we just need one (username or email) -- I will add a  2nd `break`.

mrnugget reviewed (approved):
Approve, but it would be cool to have a test for the username/email extraction logic.

Also: please make sure to _manually_ test this to make sure it does render correctly etc.

pjlast commented on client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx:
```diff
@@ -65,35 +79,63 @@ const getNormalizedAccount = (
+
+                const samlExternalData = accountExternalData as SamlExternalData
+                if (samlExternalData.Values) {
+                    const entries = Object.entries(samlExternalData.Values)
+                    for (const [name, att] of entries) {
```
You don't have to do `Object.entries(samlExternalData.Values)`, you can just directly iterate over it:

```suggestion
                    for (const [name, att] of samlExternalData.Values) {
```

This preserves type information. `Object.entries` returns `att` as an `any` type so you have no type information past that point and would have to cast it again

pjlast replied:
Since you can call `Object.entries()` on it means it should be iterable. You can read about `Object.entries()` here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/entries

But the part I want to highlight is "This is the same as iterating with a for...in loop". So I'm wondering where you are seeing that error? Can you perhaps share a screenshot?

miveronese commented on client/web/src/user/settings/auth/ExternalAccount.tsx:
```diff
@@ -52,9 +57,9 @@ export const ExternalAccount: React.FunctionComponent<React.PropsWithChildren<Pr
+                    {account.external?.userUrl ? (
                         <>
-                            {account.external.userName} (
+                            {account.external.userName}(
```
intentional by my IDE. thanks for noticing this. totally scaped my eye 🙏

miveronese commented on client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx:
```diff
@@ -45,17 +56,20 @@ interface Props {
     onDidError: (error: ErrorLike) => void
 }
 
+const emailAttrs = new Set(['http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress', 'emailaddress', '"http://schemas.xmlsoap.org/claims/EmailAddress'])
+const usernameAttrs = new Set(['nickname', 'login', 'username', 'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name'])
```
That would be ideal. The backend stores a huge tree of data in the DB that is not entirely normalized -- I am assuming that there are reasons behind that decision, so I preferred not to touch in the data structure that is stored in the DB during this PR. but here's room for improvement there, definitely

miveronese commented on client/web/src/user/settings/auth/ExternalAccount.tsx:
```diff
@@ -30,8 +30,13 @@ export const ExternalAccount: React.FunctionComponent<React.PropsWithChildren<Pr
-    }, [authProvider.authenticationURL])
+
+        if (authProvider.serviceType === 'saml') {
+            window.location.assign(`${authProvider.authenticationURL}`)
```
We have a [middleware](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/cmd/frontend/internal/auth/saml/middleware.go?L107&subtree=true) that handles GET requests to the login URL and redirects it.

If you are login in with SAML, via the COntinue with SAML button, you will be redirected to the /search page.
Here, on the Account Security page, I think it is better to keep the SAML behavior consistent with the other auth providers, meaning: after the user clicks the Add button and a SAML external_account is added, the user is redirected to the same page as it happens for GL and GH.

miveronese commented on client/web/src/user/settings/auth/ExternalAccountSignIn.test.tsx:
```diff
@@ -0,0 +1,87 @@
+        for (const testCase of testCases) {
+            expect(getUsernameOrEmail(testCase)).toEqual('adalovelace')
+        }
+    })
```
FYI: 
the test case `saml account data has no username neither email` hasn't been included.

If the saml data that comes from the IdP doesn't have username or email,  when the user clicks Add, the SAML login will fail the validation made by our backend and the user will see an error in the saml login URL. Something like this.
<img width="756" alt="Screen Shot 2022-10-26 at 12 01 24 PM" src="https://user-images.githubusercontent.com/1552863/198101942-3222cda8-43d5-47af-b743-b36f4725ad1a.png">

miveronese commented:
@kopancek , @pjlast, it's ready for your last look (hopefully the last :D)

pjlast commented on client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx:
```diff
@@ -45,17 +56,42 @@ interface Props {
 ): NormalizedMinAccount => {
     // kind and type match except for the casing
-    const kind = authProvider.serviceType.toLocaleUpperCase() as ExternalServiceKind
+    const kind = authProvider.serviceType as ExternalAccountKind
```
It might be better to use a type guard here and make sure that `.serviceType` doesn't contain any unwanted fields. Something like 
```suggestion
    ): NormalizedAccount | null => {
    if (authProvider.serviceType === 'builtin' || authProvider.serviceType === 'http-header') {
        return null
    }
```

And then

miveronese replied:
Wondering if there's a comment missing here...

pjlast commented on client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx:
```diff
@@ -65,35 +101,46 @@ const getNormalizedAccount = (
+                        userName: userNameOrEmail
+                    },
                 }
+            }
```
GitHub doesn't allow me to add a comment further down. But it might be worth changing the authProvider check, alongside the `getNormalizedAccount` type guarding I suggested above) from the following:

```
if (authProvider && authProvider.serviceType !== 'builtin') {
    const normAccount = getNormalizedAccount(accounts, authProvider)
    
    return (
```

to 

```
const normAccount = getNormalizedAccount(accounts, authProvider)
if (normAccount) {
    return (
```

pjlast replied:
This allows us to make sure that we support only the account types we want to support. For example, if you want to exclude `openidconnect` for now, you can add `|| authProvider.serviceType !== 'openidconnect'` to the type check and then it simply won't render

pjlast reviewed (commented):
One thing to keep in mind is that TypeScript only allows for compile-time safety. It can do nothing about run-time safety. So when an object comes in from some external source, you have to make sure you're annotating the objects with the correct types.

This also goes for whenever you use the `as` keyword. It's not a typecast like in Go, where it does any sort of checks. You're telling the compiler "Trust me, this is definitely this type". If, at runtime, it turns out it is not that type, you'll get runtime errors. Often you don't have to do these `as` conversions at all. Instead, it's better to check if the properties you're looking for is present, or to make sure that properties that you don't want aren't present.

kopancek commented on client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx:
```diff
@@ -26,15 +24,28 @@ interface GitLabExternalData {
 }
 
+export interface SamlExternalData {
+    Values: object
```
You are right, that using `Object.entries` basically converts any object into an iterable `Map` and that's why it works with this method. You are also right, that we should never use `object` in the type definition as that's too generic.

But I also think that specifying the type with all the possibilities for the attribute names from SAML needlessly complicates things for the UI. I mentioned it before, but I think a better way would be to transform this data on the backend side - e.g. on the graphql layer. That way, we do not need to deal with multiple keys on the UI.

However I'm fine with merging this PR as is and doing that improvement in a follow up PR.

pjlast replied:
For the current way this works, before we have the back-end doing anything, we're going to be specifying all those attribute names anyways. If it's not in the object itself, it's in separate lists. Either way, somewhere that list is going to exist, until it's reworked.

Having the list here at least allows for type inference to work as it's supposed to, instead of iterating over an opaque object's keys and values and forcing type casts with `as`

kopancek commented on client/web/src/user/settings/auth/ExternalAccountSignIn.test.tsx:
```diff
@@ -0,0 +1,87 @@
+    }
+}
+
+function samlDataObject(keysValues: object): SamlExternalData {
+    // Add some other fields to make sure we are getting the right ones
+    Object.assign(keysValues, {'any field':'false'})
+    Object.assign(keysValues, {'random field':'Mon Oct 10 2022 13:07:34 GMT+0000 (Coordinated Universal Time)'})
+    Object.assign(keysValues, {'one more random field':'banana'})
+
+    const testData: unknown = {
+        'Values': Object.fromEntries(
+            Object.entries(keysValues).map(([key, value]) =>
+                [key, toAttribute(value)]
+            )
+        )
+    }
+
+    return testData as SamlExternalData
+}
```
Can be IMO simplified to:
```suggestion
function samlDataObject(testData: object): SamlExternalData {
    // Add some other fields to make sure we are getting the right ones
    testData['any field'] = 'false'
    testData['random field'] = 'Mon Oct 10 2022 13:07:34 GMT+0000 (Coordinated Universal Time)'
    testData['one more random field'] = 'banana'

    return {
      Values: testData
    }
}
```

miveronese replied:
Tried, but got a `Element implicitly has an 'any' type because expression of type 'string' can't be used to index type` error. So will keep it how it is.

kopancek commented on client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx:
```diff
@@ -45,17 +56,42 @@ interface Props {
+        for (const [name, att] of entries) {
+            const attribute = att as Attribute
+            if (usernameAttrs.has(name)) {
+                userNameOrEmail = attribute.Values.length > 0 && attribute.Values[0].Value || ''
```
Nit
```suggestion
                userNameOrEmail = attribute.Values[0]?.Value || ''
```

miveronese replied:
This part has changed, but I added this anyway.

kopancek commented on client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx:
```diff
@@ -65,35 +101,46 @@ const getNormalizedAccount = (
+            case 'saml': {
+                const samlExternalData = accountExternalData as SamlExternalData
+                // In case the SAML values don't have a username, we get the user email.
+                const userNameOrEmail = getUsernameOrEmail(samlExternalData)
+                normalizedAccount = {
+                    ...normalizedAccount,
+                    external: {
+                        id: account.id,
+                        userName: userNameOrEmail
+                    },
```
Nit - we can inline the `userNameOrEmail` variable, since it's only used once.
```suggestion
                normalizedAccount = {
                    ...normalizedAccount,
                    external: {
                        id: account.id,
                        userName: getUsernameOrEmail(samlExternalData)
                    },
```

miveronese commented:
@kopancek  and @pjlast , I addressed your suggestions 🙇 And here is an updated manual test just in case

https://user-images.githubusercontent.com/1552863/198381703-14d22e60-7437-485c-8006-d85e252a78bc.mov

pjlast commented on client/web/src/user/settings/auth/ExternalAccountsSignIn.tsx:
```diff
@@ -26,15 +24,36 @@ interface GitLabExternalData {
+        nickname?: Attribute
+        login?: Attribute
+        'http://schemas.xmlsoap.org/ws/2005/05/identity/claims/name'?: Attribute
+        any?: Attribute
```
Just wondering what the point of this field is?

miveronese replied:
my eye failed me 😢  -- it is supposed to be `username` 🤦‍♀️

pjlast replied:
That's okay, that's what PRs are for :)

pjlast reviewed (approved):
Just wondering what the point of that `any?` field is? otherwise LGTM 🚀 :shipit:

kopancek reviewed (approved):
LGTM 🚢
